src/components/ha-area-picker.ts

```diff
@@ -0,0 +1,229 @@
+  @property() public value?: string;
+  @property() public _areas?: AreaRegistryEntry[];
+  @property({ type: Boolean, attribute: "no-add" })
+  public noAdd?: boolean;
+  @property() private _opened?: boolean;
+
+  public hassSubscribe(): UnsubscribeFunc[] {
+    return [
+      subscribeAreaRegistry(this.hass.connection!, (areas) => {
+        this._areas = this.noAdd
+          ? areas
+          : [
+              ...areas,
+              {
+                area_id: "add_new",
+                name: this.hass.localize("ui.components.area-picker.add_new"),
+              },
+            ];
+      }),
+    ];
+  }
+
+  protected render(): TemplateResult | void {
+    if (!this._areas) {
+      return;
+    }
+    return html`
+      <vaadin-combo-box-light
+        item-value-path="area_id"
+        item-id-path="area_id"
+        item-label-path="name"
+        .items=${this._areas}
+        .value=${this._value}
+        .renderer=${rowRenderer}
+        @opened-changed=${this._openedChanged}
+        @value-changed=${this._areaChanged}
+      >
+        <paper-input
+          .label=${this.label === undefined && this.hass
+            ? this.hass.localize("ui.components.area-picker.area")
+            : this.label}
+          class="input"
+          autocapitalize="none"
+          autocomplete="off"
+          autocorrect="off"
+          spellcheck="false"
+        >
+          ${this.value
+            ? html`
+                <paper-icon-button
+                  aria-label=${this.hass.localize(
+                    "ui.components.area-picker.clear"
+                  )}
+                  slot="suffix"
+                  class="clear-button"
+                  icon="hass:close"
+                  @click=${this._clearValue}
+                  no-ripple
+                >
+                  ${this.hass.localize("ui.components.area-picker.clear")}
+                </paper-icon-button>
+              `
+            : ""}
+          ${this._areas.length > 0
+            ? html`
+                <paper-icon-button
+                  aria-label=${this.hass.localize(
+                    "ui.components.area-picker.show_areas"
+                  )}
+                  slot="suffix"
+                  class="toggle-button"
+                  .icon=${this._opened ? "hass:menu-up" : "hass:menu-down"}
+                >
+                  ${this.hass.localize("ui.components.area-picker.toggle")}
+                </paper-icon-button>
+              `
+            : ""}
+        </paper-input>
+      </vaadin-combo-box-light>
+    `;
+  }
+
+  private _clearValue(ev: Event) {
+    ev.stopPropagation();
+    this._setValue("");
+  }
+
+  private get _value() {
+    return this.value || "";
+  }
+
+  private _openedChanged(ev: PolymerChangedEvent<boolean>) {
+    this._opened = ev.detail.value;
+  }
+
+  private _areaChanged(ev: PolymerChangedEvent<string>) {
+    const newValue = ev.detail.value;
+
+    if (newValue === "add_new") {
```

Comment:
FYI, if you have a large IF-block at which you return at the end, and then have a small if-block, you can better reverse them and then have less indentation. Easier to follow.

```ts
    if (newValue !== this._value) {
      this._setValue(newValue);
    }

    if (newValue !== "add_new") { return; }
```